### DISCUSSION TITLE
Working example of DriftIsolate

### DISCUSSION THREAD

**Discussion by MaxiStefan** [simolus3/drift#1769] [Q&A]
Hi,

I am trying to implement isolates on DB operations in my Flutter app and I was curious whether anyone tried implementing Drift with Flutter_Bloc as DI. I am using repositories as wrappers over db DAOs and HTTP data providers and it's kinda janky to switch to isolates inside a centralized repository.

**Comment by simolus3** [comment]
> it's kinda janky to switch to isolates inside a centralized repository.

If you provide the instance of your app's database class via some DI mechanism, none of your repositories would need to change. When you use isolates with drift, you only need to change how you open your database. All other methods are then using the background isolate under the hood, you shouldn't have to adapt anything.

I'm aware of [this example project](https://github.com/rodydavis/moor_shared) and I can contribute isolate changes to it if it helps. If you were looking at other samples I can change them to use isolates too.

But if there's anything unclear about setting up isolates in [the documentation](https://drift.simonbinder.eu/docs/advanced-features/isolates/) please let me know so that I can improve them!

**Reply by MaxiStefan** [reply]
Hi,

Thank you for replying to me! Let's consider my scenario. I have the following structure:
├── BLocDatabaseProvider(1 single db conn in a state)
│   └── Repository 1 - usess BlocProvider.of<DatabaseProvider>(context).db.TodoDao
│   └── Repository 2 - usess BlocProvider.of<DatabaseProvider>(context).db.TodoDao3
│   └── Repository n - usess BlocProvider.of<DatabaseProvider>(context).db.TodoDao2
│       └── BLocProvider feature1 - uses Repo1
│       └── BLocProvider feature2 - uses Repo2
│       └── BLocProvider feature3 - uses Repo1, Repo2 and Repo3

Feature3 needs to parse 25k +  elements and execute DB queries which takes ~10 mins on profiling and on an emulator. For this scenario, I did manage to create a db isolate but locally(within a method exec), while I changed the reference of DatabaseProvider, it did not propagate to Repo1,2,3 and I have an idea why but not yet tested.

I decide 10 mins is not swift enough and the devices the app will be running on are quite powerful, maybe octa-core processors. How can I open 2+ isolates If i have a state management like Bloc providing my db conn. I have thought of grabbing all the data in a list structure and then send it to an isolate that executes a batch update, however that throws a FK exception. And yet the single execution of all update queries do not?

Also, the documentation for the creation of the isolate is pretty solid. Complexity with it comes in different scenarios that people have. It would be helpful for others if either you or people you helped can provide some mock examples of how they integrated isolates in a more complex way( like I believe i have here :) )

**Reply by simolus3** [reply]
> however that throws a FK exception

Is this on a single background isolate adding all the entries? If you split the batch across multiple background isolates you'll need to ensure that you split it in a way that data only references rows from the same chunk.

For these large batch inserts, I wouldn't even use the `DriftIsolate` helper to synchronize data in real time. You can just spawn the worker isolate to process the data, and that one would just create its own database connection without the synchronization mechanisms provided by `DriftIsolate`. After you're done inserting data, call [`markTablesUpdated`](https://drift.simonbinder.eu/api/drift/databaseconnectionuser/marktablesupdated) on the main isolate so that stream queries can update.

**Reply by MaxiStefan** [reply]
Yes, I wanted to batch update on the UI thread initially. It is actually 50k update queries that I launch.

So I should just remove  DriftIsolate.inCurrent(
      () =>...)
and just use 
    final driftIsolate = DatabaseConnection.fromExecutor(executor) ?
    
  Edit: If i remove everything DriftIsolate related and use the Connection.FromExecutor i seem to be running it on the main thread instead. Do you have the syntax for isolate without your helper?

**Reply by simolus3** [reply]
> If i remove everything DriftIsolate related and use the Connection.FromExecutor i seem to be running it on the main thread instead

Yes, of course. For isolates, there are essentially two options that drift supports:

1. Fetch data and build queries in the main isolate, but execute them in a background isolate. This is what `DriftIsolate.spawn` is doing, and it means that most of your code runs on the main isolate still. However, queries are running in parallel on a background isolate. For almost all apps, __this is what I recommend__. It's relatively easy to set up and you don't need to change anything about how you use the database, it only affects the setup. This is what's described in the [documentation](https://drift.simonbinder.eu/docs/advanced-features/isolates/).
2. Do most work entirely in the main isolate, but manually spawn background isolates for some compute-heavy tasks (such as writing 50k rows in a batch). Just use `compute` from Flutter, and inside the callback create another instance of the database that then also runs on a background isolate (because you created it inside a `compute` callback). You don't need a `DriftIsolate` for this.

Essentially, a `DriftIsolate` is a tool to have the database implementation run on one isolate and have the usages of the database on another isolate. Dart doesn't allow sharing mutable state between isolates, so `DriftIsolate` is internally using message-passing to send queries and responses across isolates. However, you can still use drift across isolates without a `DriftIsolate` if you don't need to share the same database instance across isolates. If each isolate that you manage uses its own database connection, there's no need for `DriftIsolate`.

**Comment by MaxiStefan** [comment]
hi,

Thanks for the recommendation. I got compute to work.

However, I am trying to maximize the capacity of the CPU and concurrently access the database to update tables from, say 3 threads. I get a Database is locked -> Causing statement: PRAGMA user_version = 1; error. Do you know if what I am trying to do is achievable or whether PRAGMA Synchronziation = off works?

Thanks!
